pallets/circuit-breaker/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+		type TechnicalOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The maximum percentage of a pool's liquidity that can be traded in a block.
+		type DefaultMaxNetTradeVolumeLimitPerBlock: Get<Percent>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Default maximum net trade volume limit per block
+	#[pallet::type_value]
+	pub fn DefaultTradeVolumeLimit<T: Config>() -> Percent {
+		T::DefaultMaxNetTradeVolumeLimitPerBlock::get()
+	}
+
+	#[pallet::storage]
+	/// Allowed liquidity range per block, calculated based on the initial liquidity and trade volume limit percentage
+	#[pallet::getter(fn allowed_liqudity_range_per_asset)]
+	pub type AllowedLiquidityRangePerAsset<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, (T::Balance, T::Balance)>;
+
+	#[pallet::storage]
+	/// Trade volume limits of assets that don't use the default value
+	#[pallet::getter(fn trade_volume_limit_per_asset)]
+	pub type TradeVolumeLimitPerAsset<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, Percent, ValueQuery, DefaultTradeVolumeLimit<T>>;
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {
+		/// Allowed liquidity limit is not stored for asset
+		LiquidityLimitNotStoredForAsset,
+		/// Minimum pool trade volume per block has been reached
+		MinTradeVolumePerBlockReached,
+		/// Maximum pool trade volume per block has been reached
+		MaxTradeVolumePerBlockReached,
+		/// Invalid trade volume limit. Limit must be non-zero.
+		InvalidTradeVolumeLimit,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(<T as Config>::WeightInfo::set_trade_volume_limit())]
+		#[transactional]
+		pub fn set_trade_volume_limit(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			trade_volume_limit: Percent,
+		) -> DispatchResult {
+			T::TechnicalOrigin::ensure_origin(origin)?;
+			ensure!(!trade_volume_limit.is_zero(), Error::<T>::InvalidTradeVolumeLimit);
+			<TradeVolumeLimitPerAsset<T>>::insert(asset_id, trade_volume_limit);
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn calculate_and_store_liquidity_limits(asset_id: T::AssetId, initial_liquidity: T::Balance) -> DispatchResult {
+		if !<AllowedLiquidityRangePerAsset<T>>::contains_key(asset_id) {
+			let liquidity_diff = Pallet::<T>::trade_volume_limit_per_asset(asset_id).mul_floor(initial_liquidity);
+			let min_limit = initial_liquidity
+				.checked_sub(&liquidity_diff)
+				.ok_or(ArithmeticError::Overflow)?;
+			let max_limit = initial_liquidity
+				.checked_add(&liquidity_diff)
+				.ok_or(ArithmeticError::Overflow)?;
+			<AllowedLiquidityRangePerAsset<T>>::insert(asset_id, (min_limit, max_limit));
+		}
+		Ok(())
+	}
+
+	fn test_liquidity_limits(asset_id: T::AssetId, updated_liquidity: T::Balance) -> DispatchResult {
```

Comment:
i would prefer something else than "test" . 

because i was looking at this like " why a test is here" first. 

eg.

```suggestion
	fn ensure_liquidity_limits(asset_id: T::AssetId, updated_liquidity: T::Balance) -> DispatchResult {
```